File: apps/mobile/ios/Sources/FlagshipAPI/Models/BrowserStreamModels.swift
import Foundation

// P8 — browser-viewer wire format.
//
// MIRRORS apps/web/public/webapp/views/browser-viewer.js byte-for-byte.
// The daemon's WS path is `/api/screens/browser-tabs/:tabId/stream`
// (P1.11); the session token rides the URL query param `sessionToken`
// because URLSessionWebSocketTask + OkHttp can't add a custom header
// before the upgrade handshake on every platform.

public enum BrowserFrame: Equatable, Sendable {
    case frame(dataBase64: String)
    case error(message: String)

    private enum CodingKeys: String, CodingKey {
        case kind, dataBase64, message
    }

    public static func decode(_ json: Data) -> BrowserFrame? {
        guard let obj = try? JSONSerialization.jsonObject(with: json) as? [String: Any] else { return nil }
        let kind = obj["kind"] as? String
        switch kind {
        case "frame":
            guard let s = obj["dataBase64"] as? String else { return nil }
            return .frame(dataBase64: s)
        case "error":
            let msg = (obj["message"] as? String) ?? "stream error"
            return .error(message: msg)
        default:
            return nil
        }
    }
}

public enum BrowserInput: Equatable, Sendable {
    case mouseDown(x: Int, y: Int, button: String)
    case mouseUp(x: Int, y: Int, button: String)
    case mouseMove(x: Int, y: Int)
    case scroll(x: Int, y: Int, deltaX: Double, deltaY: Double)
    case key(eventType: String, key: String, code: String)

    public func wireDictionary() -> [String: Any] {
        let inputPayload: [String: Any]
        switch self {
        case .mouseDown(let x, let y, let button):
            inputPayload = ["kind": "mouseDown", "x": x, "y": y, "button": button]
        case .mouseUp(let x, let y, let button):
            inputPayload = ["kind": "mouseUp", "x": x, "y": y, "button": button]
        case .mouseMove(let x, let y):
            inputPayload = ["kind": "mouseMove", "x": x, "y": y]
        case .scroll(let x, let y, let dx, let dy):
            inputPayload = ["kind": "scroll", "x": x, "y": y, "deltaX": dx, "deltaY": dy]
        case .key(let evt, let key, let code):
            inputPayload = ["kind": "key", "eventType": evt, "key": key, "code": code]
        }
        return ["kind": "input", "input": inputPayload]
    }

    public func encode() throws -> Data {
        try JSONSerialization.data(withJSONObject: wireDictionary(), options: [.sortedKeys])
    }
}

File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/api/BrowserStreamModels.kt
// P8 — browser-viewer wire format.
//
// MIRRORS apps/web/public/webapp/views/browser-viewer.js byte-for-byte
// AND apps/mobile/ios/Sources/FlagshipAPI/Models/BrowserStreamModels.swift.
// The daemon's WS path is `/api/screens/browser-tabs/:tabId/stream`
// (P1.11); the session token rides the URL query param `sessionToken`
// because OkHttp's `WebSocket` upgrade reuses the connection's headers
// once but the daemon-side parser keys auth off the query param the
// same way the webapp does.

package com.flagshipserver.app.api

import org.json.JSONObject

sealed interface BrowserFrame {
    data class Frame(val dataBase64: String) : BrowserFrame
    data class Err(val message: String) : BrowserFrame

    companion object {
        fun decode(json: String): BrowserFrame? = try {
            val obj = JSONObject(json)
            when (obj.optString("kind")) {
                "frame" -> {
                    val b64 = obj.optString("dataBase64", "")
                    if (b64.isEmpty()) null else Frame(b64)
                }
                "error" -> Err(obj.optString("message", "stream error"))
                else -> null
            }
        } catch (_: Throwable) {
            null
        }
    }
}

sealed interface BrowserInput {
    data class MouseDown(val x: Int, val y: Int, val button: String) : BrowserInput
    data class MouseUp(val x: Int, val y: Int, val button: String) : BrowserInput
    data class MouseMove(val x: Int, val y: Int) : BrowserInput
    data class Scroll(val x: Int, val y: Int, val deltaX: Double, val deltaY: Double) : BrowserInput
    data class Key(val eventType: String, val key: String, val code: String) : BrowserInput

    fun toWireJson(): String {
        val outer = JSONObject()
        outer.put("kind", "input")
        val inner = JSONObject()
        when (this) {
            is MouseDown -> {
                inner.put("kind", "mouseDown"); inner.put("x", x); inner.put("y", y); inner.put("button", button)
            }
            is MouseUp -> {
                inner.put("kind", "mouseUp"); inner.put("x", x); inner.put("y", y); inner.put("button", button)
            }
            is MouseMove -> {
                inner.put("kind", "mouseMove"); inner.put("x", x); inner.put("y", y)
            }
            is Scroll -> {
                inner.put("kind", "scroll"); inner.put("x", x); inner.put("y", y)
                inner.put("deltaX", deltaX); inner.put("deltaY", deltaY)
            }
            is Key -> {
                inner.put("kind", "key"); inner.put("eventType", eventType)
                inner.put("key", key); inner.put("code", code)
            }
        }
        outer.put("input", inner)
        return outer.toString()
    }
}
